ui-lib/src/main/java/co/electriccoin/zcash/ui/common/serialization/metada/MetadataKey.kt

```diff
@@ -16,66 +17,99 @@ import com.google.crypto.tink.util.SecretBytes
                 METADATA_ENCRYPTION_KEY_SIZE
             )
         return ChaCha20Poly1305Key.create(SecretBytes.copyFrom(subKey, access))
     }
 
+    fun deriveFirstDecryptionKey(salt: ByteArray): ChaCha20Poly1305Key {
+        return deriveDecryptionkey(salt, encryptionBytes, "encryption_key")
+    }
+
+    fun deriveSecondDecryptionKey(salt: ByteArray): ChaCha20Poly1305Key? {
+        if (decryptionBytes == null) return null
+        return deriveDecryptionkey(salt, decryptionBytes, "decryption_key")
+    }
+
+    private fun deriveDecryptionkey(
+        salt: ByteArray,
+        decryptionBytes: SecretBytes,
+        infoKey: String
+    ): ChaCha20Poly1305Key {
+        assert(salt.size == METADATA_SALT_SIZE)
+        val access = InsecureSecretKeyAccess.get()
+        val subKey =
+            Hkdf.computeHkdf(
+                "HMACSHA256",
+                decryptionBytes.toByteArray(access),
+                null,
+                salt + infoKey.toByteArray(),
+                METADATA_ENCRYPTION_KEY_SIZE
+            )
+        return ChaCha20Poly1305Key.create(SecretBytes.copyFrom(subKey, access))
+    }
+
     companion object {
-        /**
-         * Derives the long-term key that can decrypt the given account's encrypted
-         * address book.
-         *
-         * This requires access to the seed phrase. If the app has separate access
-         * control requirements for the seed phrase and the address book, this key
-         * should be cached in the app's keystore.
-         */
         suspend fun derive(
             seedPhrase: SeedPhrase,
             network: ZcashNetwork,
-            account: WalletAccount
+            zashiAccount: ZashiAccount,
+            selectedAccount: WalletAccount
         ): MetadataKey {
             val key =
-                DerivationTool.getInstance().deriveArbitraryAccountKey(
-                    contextString = "ZashiMetadataEncryptionV1".toByteArray(),
-                    seed = seedPhrase.toByteArray(),
-                    network = network,
-                    accountIndex = account.hdAccountIndex,
-                )
-            return MetadataKey(SecretBytes.copyFrom(key, InsecureSecretKeyAccess.get()))
+                DerivationTool.getInstance()
+                    .deriveAccountMetadataKey(
+                        seed = seedPhrase.toByteArray(),
+                        network = network,
+                        accountIndex = zashiAccount.hdAccountIndex,
+                    )
+                    .derivePrivateUseMetadataKey(
+                        ufvk =
+                            when (selectedAccount) {
+                                is KeystoneAccount -> selectedAccount.sdkAccount.ufvk
+                                is ZashiAccount -> null
+                            },
+                        network = network,
+                        privateUseSubject = "metadata".toByteArray()
+                    )
```

Comment:
Bug: `"ZashiMetadataEncryptionV1"` should have been kept as the `privateUseSubject` (for the same reason it was being used as the `contextString`. At a minimum it should have included `Zashi` as it's for Zashi's private use.